### ISSUE TITLE
3D Text scale & translation issues when not fullscreen or billboard

### ISSUE THREAD

**Issue by phr00t** [xenko3d/xenko#314]
**Release Type**: Official Release & Github

**Version**: 3.1 beta

**Platform(s)**: Windows 10 x64

**Describe the bug**
Adding a UI element to the scene, with a TextBlock that is not a fullscreen billboard, doesn't get scaled & translated correctly. It will deform, stretch and rotate with the camera in unexpected ways.

**To Reproduce**
1. Add a UI element to the scene
2. Make the UI element use a page with just a TextBlock, type "Hello World" into it
3. Uncheck the "fullscreen" & "billboard" features of the UI
4. Move and rotate the camera around & observe the TextBlock

**Expected behavior**
The TextBlock should continue to point in a certain direction at an expected size.

**Comment by phr00t** [comment]
Suspecting the bug is around here somewhere:

https://github.com/xenko3d/xenko/blob/master/sources/engine/Xenko.Graphics/UIBatch.cs#L437

Unsure where the "fullscreen" boolean is handled -- since TextBlock works when it is fullscreen... so we don't want behavior to change if fullscreen is selected.

**Comment by phr00t** [comment]
Forcing drawCommand.RealVirtualResolutionRatio = Vector2.One; significantly improves this, but doesn't solve it alone. If I force it to Vector2.One when going into InternalUIDraw, the text scale seems correct in 3D. It may also solve the flickering. However, it still gets translated incorrectly -- the text rotates almost as if in the inverse direction it should be (guess)...

**Comment by phr00t** [comment]
I highly suspect I've discovered the problem!

There are two types of text rendering functions: SpriteBatch.DrawString & UIBatch.DrawString. It looks like TextBlocks do not use SpriteBatch.DrawString when in 3D "world space" like they should -- they always use UIBatch.DrawString and assume it is a 2D overlay.

I haven't confirmed this, because TextBlocks setup a SpriteFont.InternalUIDrawCommand for its UIBatch to draw... when a SpriteFont.InternalDrawCommand is needed for a SpriteBatch to handle (but I can't find a SpriteBatch reference...).

Basically, it looks like a TextBlock (and all other UI or 3D texts) need a UIBatch & SpriteBatch, and use the right type whether it is in 3D (e.g. not a billboard) or 2D (billboard/fullscreen overlay).

**Comment by phr00t** [comment]
Maybe not... SpriteBatch only takes in a Vector2 as a position, so definitely not 3D....

**Comment by phr00t** [comment]
Testing commits made by xen2 (on my fork, so please confirm bug on master) show they work on left justified/aligned 3d text, but not for center or right aligned/justified. Center or right justified text slides horizontally.
